### PR TITLE
Refactor existing JIT testing utils to enable new OpInfo test suite to reuse existing logic

### LINKED ISSUE #44506: Support OpInfo-based JIT testing
Today in PyTorch test_autograd.py and test_jit.py use "method tests" from common_methods_invocations.py to generate tests. Pointer:

https://github.com/pytorch/pytorch/blob/4bead6438a4271f154980dc242199c2f1b897bc1/torch/testing/_internal/common_methods_invocations.py#L364

These tuples are tricky to use and hard to maintain, however. To facilitate more maintainable and thorough tests we're switching the test suite to use OpInfos, instead. Pointer:

https://github.com/pytorch/pytorch/blob/4bead6438a4271f154980dc242199c2f1b897bc1/torch/testing/_internal/common_methods_invocations.py#L201

OpInfos have significant advantages vs. the method test tuples:

- an OpInfo is a class, which makes it easy to add and query operator properties
- OpInfos are intended for use with multiple tests
- the device generic testing framework supports OpInfos

To switch from method_tests to OpInfos, however, tests using method_tests have to be updated to use OpInfos.@albanD and I already reviewed test_autograd.py's use of method_tests and developed better, more comprehensive, and more readable tests that use OpInfos instead. Compare the method_test-based tests here:

https://github.com/pytorch/pytorch/blob/4bead6438a4271f154980dc242199c2f1b897bc1/test/test_autograd.py#L4669

with the new OpInfo based grad and gradgrad checks here:

https://github.com/pytorch/pytorch/blob/4bead6438a4271f154980dc242199c2f1b897bc1/test/test_ops.py#L94

I expect similar results from revising the generated tests in test_jit.py, and, as mentioned above, this revision is necessary for us to move away from method_tests. For test_autograd.py this revision process took ~4 hours. 

cc @gmagogsfm @mruberry @VitalyFedyunin @jansel

### PR DESCRIPTION
**Pull request by lobstergrindset**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48627 Refactored OpInfo testing to support custom SampleInputs, added addmm to op_db to test
* #47696 Support for OpInfo-based testing for operators in JIT
* #48567 Refactored assertAutodiff test to have better error message
* **#47695 Refactor existing JIT testing utils to enable new OpInfo test suite to reuse existing logic**

The method_tests from common_methods_invoations.py are being migrated into a new OpInfo class-based testing framework. The work in this commit pulls out the functions embedded in the old method_tests logic and places them in a location that both the old method_tests and OpInfo tests can use

Specifically: created torch/testing/_internal/common_jit.py from functions and methods in torch/testing/_internal/jit_utils.py and test/test_jit.py. Also created new intermediate class JitCommonTestCase to house moved methods. Also slightly modified jit_metaprogramming_utils.py to work for OpInfo tests

Differential Revision: [D25212437](https://our.internmc.facebook.com/intern/diff/D25212437)

### CODE REVIEW

File: torch/testing/_internal/jit_metaprogramming_utils.py [hunk 1]
```diff
@@ -243,6 +243,8 @@ def get_call(method_name, func_type, args, kwargs):
         call = '{}.{}({})'.format(self_arg, method_name, argument_str)
     elif func_type == 'nn_functional':
         call = 'torch.nn.functional.{}({})'.format(method_name, argument_str)
+    elif func_type == 'function':
```
Comment (mruberry) [inline torch/testing/_internal/jit_metaprogramming_utils.py:246]: function vs functional?

**Comment by mruberry** [comment]
This looks good to me, @eellison would you like to look at it from the JIT perspective?

File: torch/testing/_internal/common_jit.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            or all([any(g.findNode(n) is not None for g in diff_subgraphs) for n in nonfusible_nodes])
+
+        # For any fusible node, it must show up in one of the FusionGroup in the DifferentiableGraph.
+        fusion_nodes = list(chain.from_iterable([g.findAllNodes('prim::FusionGroup') for g in diff_subgraphs]))
+        fusion_subgraphs = [node.g('Subgraph') for node in fusion_nodes]
+        found_all_fusible_nodes = (len(fusion_nodes) == 0 and len(fusible_nodes) == 0)\
+            or all([any(g.findNode(n) is not None for g in fusion_subgraphs) for n in fusible_nodes])
+
+        err_msg = "Failure in testing node's autodifferentiation, "
+        if should_autodiff_node:
```
Comment (mruberry) [inline torch/testing/_internal/common_jit.py:161]: We need to expand these error messages so they enumerate ALL the failures. That is, "found node X that should be fused but was not, found node Y that was fused that shouldn't be..." and suggest the appropriate remedy, "Add X to the list of fusible nodes..., Add Y to the list of nonfusible nodes..."

File: torch/testing/_internal/common_jit.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            err_msg += "one or more nodes were expected to be autodiffed, " \
+                "but weren't found in specified fusible/nonfusible DifferentiableGraph group"
+        else: 
+            err_msg += "one or more nodes weren't expected to be autodiffed, " \
```
Comment (eellison) [inline torch/testing/_internal/common_jit.py:165]: These error messages are much better! in the future, IMO it's nice to separate out mechanical movement of code from real changes

**Review by eellison — COMMENTED** [review]
lgtm - seems like it's mostly just code movement

File: torch/testing/_internal/common_jit.py (hunk 3 above)
Comment (Lilyjjo) [reply torch/testing/_internal/common_jit.py]: Good suggestion, I added the error message changes to a diff on top of this one

File: torch/testing/_internal/common_jit.py (hunk 2 above)
Comment (Lilyjjo) [reply torch/testing/_internal/common_jit.py]: Added specific error messages to the commit above this one in the stack

**Review by mruberry — APPROVED** [review]
Looks good to me. ping @eellison since this is a refactor of jit test code.

**Pull request closed without merging**